content/blog/multimodal-nutrition-agent/index.md

```diff
@@ -0,0 +1,340 @@
+
+template = """{% for document in documents %}
+Image: <|image_
+This image shows: {{ document.content }}
+{% endfor %}
+"""
+
+retrieval_pipeline = Pipeline()
+retrieval_pipeline.add_component("embedder", SentenceTransformersTextEmbedder(model="sentence-transformers/all-MiniLM-L6-v2"))
+retrieval_pipeline.add_component("retriever", InMemoryEmbeddingRetriever(document_store=document_store, top_k=1))
+retrieval_pipeline.add_component("prompt_builder", MultiModalPromptBuilder(template=template))
+
+retrieval_pipeline.connect("embedder.embedding", "retriever.query_embedding")
+retrieval_pipeline.connect("retriever", "prompt_builder.documents")
+```
+
+In this pipeline, the `MultiModalPromptBuilder` component receives one Document object from the retriever and renders the prompt. Notice that we have the "<|image_" placeholder in the prompt template for our model, in order to inject the images into later. Additionally, `MultiModalPromptBuilder` converts the given image into a [base64](https://en.wikipedia.org/wiki/Base64) string for the image to be processed by the multimodal agent. Let’s run the pipeline to see its output. 
+
+```python
+retrieval_pipeline.run({"embedder":{"text": "Protein bar"}})
+
+"""
+{'prompt_builder': {'prompt': '\nImage: <|image_\nThis image shows: Protein bar with chocolate peanut butter nutrition facts per bar (50g)\n',
+ 'images': ['/9j/4AAQSkZJRgABAQAAAQABAAD/4....']} 
+"""
+```
+
+## Creating the Multimodal ReAct Agent
+
+### Defining a Tool
+
+With our retrieval pipeline ready, we can create our Tool using the [DocWithImageHaystackQueryTool](https://github.com/IntelLabs/fastRAG/blob/4f73883ca85abf702eb81fd4a384a872b9fd3767/fastrag/agents/tools/tools.py#L69) component from fastRAG. `DocWithImageHaystackQueryTool` can use Haystack v2 pipelines as a tool with fastRAG Agents.
+
+This tool, like other agent tools, requires a name and a description of its functionality for our agent to decide when to use it. We provide it with our `retrieval_pipeline` as follows:
+
+```python
+from fastrag.agents.tools.tools import DocWithImageHaystackQueryTool
+
+nutrition_tool = DocWithImageHaystackQueryTool(
+    name="nutrition_tool",
+    description="useful for when you need to retrieve nutrition fact image of packaged food. It can give information about one food type per query. Pass the food name as input",
+    pipeline_or_yaml_file=retrieval_pipeline
+)
+```
+
+Let's test out our tool!
+
+```python
+tool_result = nutrition_tool.run("protein bar")
+print(tool_result[0])
+
+# Image: <|image_
+# This image shows: Protein bar with chocolate peanut butter nutrition facts per bar (50g)
+```
+
+With our tool ready, we can create our agent. 
+
+### Multimodal LLaVA Models
+
+To process both text and images, we utilize a multimodal LLM introduced in [LLaVA](https://arxiv.org/abs/2304.08485). These models combine an LLM effective in language-focused instruction-tuning with a pre-trained visual encoder skilled in visual understanding.
```

Comment:
I don't understand the relation between the used model and LLaVA. Can you verifiy that this sentence is appropriate?

Reply:
Removed the section